edgelet/aziot-edged/src/workload_manager.rs

```diff
@@ -41,6 +41,11 @@ where
         let service = edgelet_http_workload::Service::new(settings, runtime, device_info)
             .map_err(|err| EdgedError::from_err("Invalid service endpoint", err))?;
 
+        service
```

Comment:
We don't want to do this. We intentionally don't want to create / issue an Edge CA unless it's needed.

Reply:
The pre-refactor edgelet code was doing this though:
https://github.com/Azure/iotedge/blob/f147f12881c91c029de49c1aba8f509d6c3eed47/edgelet/edgelet-http-workload/src/server/mod.rs#L57-L65

Also, since we have the aziot-edged-trust-bundle contain aziot-edged-ca, aziot-edged-ca must exist upon edgeAgent's startup or the edgeAgent will fail to get the trust bundle and crash. This was the issue I was seeing.

Reply:
>The pre-refactor edgelet code was doing this though:

Then it was wrong pre-refactor too.

>Also, since we have the aziot-edged-trust-bundle contain aziot-edged-ca, aziot-edged-ca must exist upon edgeAgent's startup or the edgeAgent will fail to get the trust bundle and crash

Getting aziot-edged-trust-bundle should not fail if aziot-edged-ca doesn't exist. I assume you mean that certd's get-cert returns an error if one of the constituent certs doesn't exist. If so, that should be fixed in certd; it should ignore missing certs when merging bundles.

Reply:
Even though it's wrong, fixing it is beyond the scope of a tokio refactor. IMO the refactor should not change the behavior of edged. That would cause different behavior of other components (edgeAgent in this case) pre- and post-refactor.

As an aside, I had to keep other wrong behavior for the same reason, e.g. https://github.com/Azure/iotedge/blob/e0bcb0445b407990abfb32e8994e52280761d1e0/edgelet/edgelet-http-mgmt/src/identity/create_or_list.rs#L54-L57

Reply:
>Even though it's wrong, fixing it is beyond the scope of a tokio refactor.

The refactor of edged was already merged. This PR is not part of the refactor.

>That would cause different behavior of other components (edgeAgent in this case) pre- and post-refactor.

What different behavior does EA have with regard to the trust bundle?

Reply:
It's fixing a bug that was caused by the refactor. Pre-refactor edged always created a new Edge CA cert when it started the workload API, which the post-refactor edged omitted.

When EA starts up, the first thing it does is make a request for the trust bundle. By default, we configure the trust bundle to contain the Edge CA cert.

Since the post-refactor edged doesn't automatically create the Edge CA cert, the trust bundle request fails and EA crashes. EA then goes into a restart loop as it repeatedly restarts, tries to get the trust bundle, and fails.

We could fix it so that if the Edge CA cert doesn't exist, edged returns an empty trust bundle, as you mentioned. But AFAIK, EA only requests the trust bundle upon startup. If the Edge CA cert gets created later (for example, some module wants a server cert, which causes edged to create the Edge CA cert to issue the server cert), then how would EA know it needs to re-request the trust bundle so it contains the newly-created Edge CA cert? Would this be a problem?

Reply:
>then how would EA know it needs to re-request the trust bundle so it contains the newly-created Edge CA cert? Would this be a problem?

Why does EA need to re-request the trust bundle? Does EA ever connect to anything that uses the Edge CA?

Reply:
That's a good point. I assumed that EA was using the Edge CA cert for something if it requests the trust bundle, but I'm not sure it does.

Reply:
I tried to see if it's possible to return no trust bundle if the Edge CA cert doesn't exist. Unfortunately, that doesn't seem possible with the current response schema:

```json
{
    "certificate": "string"
}
```

i.e. EA always expects the trust bundle to be something. When empty, it fails with:
```
System.InvalidOperationException: Trusted certificates can not be null or empty.
```

So should we change EA? But that brings up compatibility issues between different versions of EA.

Reply:
Ugh, now that you mention it, that does sound familiar. I guess that's why we made pre-refactor edged do that.